tests/unit_tests/verRctNonSemanticsSimple.cpp

```diff
@@ -0,0 +1,220 @@
+        {"a2421eae8bb256644b34feeab48c6086c2c9feb40d2643436dc45e303eee8ab2", "787823abffa988b56d4a7b4a834630f71520220fd82fad035955e616ec095788"},
+        {"17d8d8dc1e1c25b7295f2eab44c4ccc08a629b8e8d781bbb6f9a51a9561bcd4c", "db1ea24be6947e03176a297160dba16d65f37751bb0ef2ba71a4590d12b61dfc"},
+        {"2c39348a9ab04dbabe3b5249819b7845ed8aaebd0d8eddd98bda0bf40753a398", "4e6cd25fbd10e2e040be84e3bf8043c612daeef625e66a5e5bcff88c9c46e82c"},
+        {"c4c97157f23b45c7084526aaa9958fe858bebe446a7efa22c491c439b74271b1", "e251db2c86193a11a5bffefffe48c20e3d92a8dc98cb3a2f41704e565bcd860a"},
+        {"d342045525139a8551bcdfa7aa0117d2ac2327cb6cf449ca59420c300e4471a5", "789c11f72060ad80f4cda5d89b24d49f9435bf765598dea7a91776e99f05f87c"},
+        {"9a972ccf2c74f648070b0be839749c98eca87166de401a6c1f59e64b938a46c1", "5444cbed5cec31fb6ed1612f815d292f2bf3d2ff584bbcd8e5201ec59670d414"},
+        {"49ccb806ccf5cbd74bae8d9fb2da8918ab61d0774ee6a6c3a6ccd237db22a088", "0c5db942fb44f29f6ef956e24db91f98a6de6e7288b0b04d01b8f260453d1431"},
+        {"74417e8d1483df2df6fe68c88fc9a72639c35d765b38351b838521addf45dadc", "a1a606d6c4762ef51c1759bcb8b5c88be1d323025400c41fe6885431064b64dc"},
+        {"48c4c349adaf7b3be27656ea70d1c83b93e1511bb0aac987861a4da9689b0e95", "ad14ffd5edac199ea7c5437d558089b0f2f03aa74bde43611322d769968b5a1c"},
+        {"2d2ffade0f85ddd83a036469e49542e93cad94f9bea535f0ea2eb2f56304517e", "bcc48d00bd06dc5439200e749d0caf8a062b072d0c0eb1f78f6a4d8f2373e5f4"},
+        {"4ee857d0ce17f66eca9c81eb326e404ceb50c8198248f2f827c440ee7aa0c0d7", "a8a9d61d4abbfb123630ffd214c834cc45113eaa51dd2f904cc6ae0c3c5d70e3"}
+    };
+
+    std::string rct_sig_to_bin_hex_string(const rct::rctSig& rs)
+    {
+        std::stringstream ss;
+        binary_archive<true> ar(ss);
+        ::do_serialize(ar, const_cast<rct::rctSig&>(rs));
+        const std::string bin_str = ss.str();
+        const unsigned char* bin_str_u8p = reinterpret_cast<const unsigned char*>(bin_str.data());
+        return epee::to_hex::string({bin_str_u8p, bin_str.size()});
+    }
+
+    rct::ctkeyM construct_original_input_pubkeys()
+    {
+        rct::ctkeyV ring0;
+
+        for (const auto& key_info : tx_input_ctkey_info)
+        {
+            rct::key dest;
+            rct::key mask;
+
+            CHECK_AND_ASSERT_THROW_MES
+            (
+                epee::from_hex::to_buffer(epee::as_mut_byte_span(dest), key_info.first),
+                "decoding dest from hex did not work"
+            );
+
+            CHECK_AND_ASSERT_THROW_MES
+            (
+                epee::from_hex::to_buffer(epee::as_mut_byte_span(mask), key_info.second),
+                "decoding mask from hex did not work"
+            );
+
+            ring0.push_back({dest, mask});
+        }
+
+        return {ring0};
+    }
+
+    const cryptonote::transaction& get_original_transaction()
```

Comment:
Does this need to be a pseudo-singleton? Note that this design locks you into a specific filename. Would be simpler and more future-proof to write a function that takes the file name for the transaction and returns the transaction. And separate functions that take a transaction and return the signatures and signatures hex. At most the additional cost is doing all this once_flag stuff in each `TEST()` case instead of once globally.

Reply:
`expand_transaction_from_bin_file_and_pubkeys` should be the change you're looking for. No more singleton, what do you think?

Reply:
This is better thanks